docs: metrics (#3121)
bentoml/BentoML: docs/source/reference/metrics.rst

===========
Metrics API
===========

> BentoML provides metrics API via `Prometheus <https://prometheus.io/>`_. Prometheus
> provides a `Python Client <https://github.com/prometheus/client_python>`_ for
> instrumenting with Prometheus.

Reviewer: ```suggestion
BentoML provides metrics API that uses `Prometheus <https://prometheus.io/>`_ under the hood.
```

> BentoML provides a ``bentoml.metrics`` API as a drop-in replacement for ``prometheus_client``:

Reviewer: ```suggestion
BentoML's ``bentoml.metrics`` is a drop-in replacement for ``prometheus_client`` that should be used in BentoML services:
```

.. code-block:: diff

   diff --git a/service.py b/service.py
   index acd8467e..0f3e6e77 100644
   --- a/service.py
   +++ b/service.py
   @@ -1,11 +1,10 @@
   -from prometheus_client import Summary
   +from bentoml.metrics import Summary
    import random
    import time

   REQUEST_TIME = Summary("request_processing_seconds", "Time spent processing request")

   @REQUEST_TIME.time()
   def process_request(t):
       """A function that takes some time."""

> Difference between ``bentoml.metrics`` and ``prometheus_client``
> ----------------------------------------------------------------

Reviewer: I think this section should be moved above the line and examples above or just removed

```suggestion
```

> While ``bentoml.metrics`` contains all API that is offered by ``prometheus_client``, there are a few differences between the two:

Reviewer: I think this is overexplained; I think we should just tell users why they need to use `bentoml.metrics` instead of a prometheus client.

* All metrics from ``bentoml.metrics`` will set up ``registry`` to handle multiprocess
  mode, which means that when setting up metrics, you **SHOULD NOT** pass in ``registry``:

  .. code-block:: python
     :caption: service.py

     # THIS WILL NOT WORK
     from bentoml.metrics import Summary, CollectorRegistry
     from bentoml.metrics import multiprocess

     registry = CollectorRegistry()
     multiprocess.MultiProcessCollector(registry)
     REQUEST_TIME = Summary(
        "request_processing_seconds", "Time spent processing request", registry=registry
     )

  instead:

  .. code-block:: python
     :caption: service.py

     # THIS WILL WORK
     from bentoml.metrics import Summary

     REQUEST_TIME = Summary("request_processing_seconds", "Time spent processing request")

Since ``bentoml.metrics`` extends ``prometheus_client``, all of the limitations from ``prometheus_client`` will 
also apply to ``bentoml.metrics``. See `limitation notes <https://github.com/prometheus/client_python#multiprocess-mode-eg-gunicorn>`_ for more details.

-----

> The following section will go over the most commonly used metrics API in
> ``bentoml.metrics``:

Reviewer: ```suggestion
``bentoml.metrics`` API reference:
```
